src/cli/net/client/mod.rs

```diff
@@ -38,25 +48,63 @@ impl MpcHandle for MpcHttpConnection {
 #[allow(dead_code)]
 impl MpcHttpConnection {
     #[must_use]
-    pub fn new(addr: &str) -> Self {
+    pub fn new(addr: Uri) -> Self {
         // this works for both http and https
         let https = HttpsConnector::new();
         let client = Client::builder().build::<_, Body>(https);
 
         Self {
             client,
-            addr: addr.parse().expect("Cannot parse the URI"),
+            addr_parts: addr.into_parts(),
         }
     }
 
+    /// same as new, but first parses the addr from a [&str]
+    /// # Errors
+    /// if addr is an invalid [Uri], this will fail
+    pub fn with_str_addr(addr: &str) -> Result<Self, InvalidUri> {
+        addr.parse().map(Self::new)
+    }
+
+    fn build_uri<T>(&self, p_and_q: T) -> Result<Uri, MpcClientError>
+    where
+        PathAndQuery: TryFrom<T>,
+        <PathAndQuery as TryFrom<T>>::Error: Into<axum::http::Error>,
+    {
+        Ok(uri::Builder::new()
+            .scheme(self.addr_parts.scheme.as_ref().unwrap().clone())
+            .authority(self.addr_parts.authority.as_ref().unwrap().clone())
+            .path_and_query(p_and_q)
+            .build()?)
+    }
+
     async fn echo(&self, s: &str) -> Result<Vec<u8>, MpcClientError> {
-        let uri: Uri = format!("{}echo?foo={}", self.addr, s)
-            .parse()
-            .expect("Failed to build an URI for \"echo\" command");
+        let uri = self.build_uri(format!("/echo?foo={}", s))?;
 
         let response = self.client.get(uri).await?;
         let result = hyper::body::to_bytes(response.into_body()).await?;
 
         Ok(result.to_vec())
     }
+
+    async fn mul<S: Step, M: Message>(
+        &self,
+        query_id: QueryId,
+        step: S,
+        messages: &[MessageEnvelope<M>],
```

Comment:
I am not sure this client API is useful for us. as a client I want to multiply `a`\*`b`. `/mul/query-id/{}/step/{}` is an endpoint for helpers to send intermediate results of multiplication before they can produce some result. as such, I think we should provide more high level API to multiply two values. For each helper client, that would mean taking two secret shares (a and b) and sending back a secret share (a*b). 

We should also keep in mind that we don't want this API to be available in prod because it reveals the result of multiplication

Reply:
I'm curious -- what will the mechanism be to communicate the intermediate results of the multiplication then?

Reply:
I agree that helper endpoint should accept `mul/query-id/...` requests, so server implementation for this endpoint is important. I would not expect any client application to call this API directly though

Reply:
eh, I guess I missed that change from the refactoring PR: https://github.com/martinthomson/raw-ipa/tree/9adbe555be149c0b3e1cc3b0afd9f6e72b6181a6/src/net. 

`net` module does not belong to `cli`, it should reside in  the `src` folder. would you be able to move it back as part of this change? I guess that removes the confusion about the client API as well